From abf72bdddfa61a4a90ae0edea921d5d2cd106f89 Mon Sep 17 00:00:00 2001
From: Tian Yang <tiany@codeaurora.org>
Date: Wed, 29 Jul 2020 17:39:28 -0700
Subject: [PATCH] skb_recycler: Add a cpustate for skb_recycler

Add one cpu hotplug state for skb_recycler, called CPUHP_SKB_RECYCLER_DEAD
to avoid using NET_DEV state, this solves a warning calltrace issue from net_dev
since it cannot register its NET_DEV_CPU_DEAD callback during its initialization.

Signed-off-by: Tian Yang <tiany@codeaurora.org>
Change-Id: I6f5729ee300248ade42317114847959fda42dd20
---
 include/linux/cpuhotplug.h | 1 +
 net/core/skbuff_recycle.c  | 6 +++---
 2 files changed, 4 insertions(+), 3 deletions(-)

--- a/include/linux/cpuhotplug.h
+++ b/include/linux/cpuhotplug.h
@@ -94,6 +94,7 @@ enum cpuhp_state {
 	CPUHP_RADIX_DEAD,
 	CPUHP_PAGE_ALLOC,
 	CPUHP_NET_DEV_DEAD,
+	CPUHP_SKB_RECYCLER_DEAD,
 	CPUHP_PCI_XGENE_DEAD,
 	CPUHP_IOMMU_IOVA_DEAD,
 	CPUHP_LUSTRE_CFS_DEAD,
--- a/net/core/skbuff_recycle.c
+++ b/net/core/skbuff_recycle.c
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2013-2016, 2019, The Linux Foundation. All rights reserved.
+ * Copyright (c) 2013-2016, 2019-2020, The Linux Foundation. All rights reserved.
  *
  * Permission to use, copy, modify, and/or distribute this software for any
  * purpose with or without fee is hereby granted, provided that the above
@@ -248,7 +248,7 @@ static int skb_cpu_callback(unsigned int
 	spin_unlock(&glob_recycler.lock);
 #endif
 
-	return NOTIFY_OK;
+	return NOTIFY_DONE;
 }
 
 #ifdef CONFIG_SKB_RECYCLER_PREALLOC
@@ -532,7 +532,7 @@ void __init skb_recycler_init(void)
 	if (skb_prealloc_init_list())
 		pr_err("Failed to preallocate SKBs for recycle list\n");
 #endif
-	cpuhp_setup_state_nocalls(CPUHP_NET_DEV_DEAD, "net/skbuff_recycler:dead:",NULL, skb_cpu_callback);
+	cpuhp_setup_state_nocalls(CPUHP_SKB_RECYCLER_DEAD, "net/skbuff_recycler:dead:",NULL, skb_cpu_callback);
 	skbuff_debugobj_register_callback();
 	skb_recycler_init_procfs();
 }
